src/backtrace.c

```diff
@@ -0,0 +1,682 @@
+ * the entry is returned. If there are multiple entries with the same address, the first
+ * entry is returned (this is the case for inlined functions: so some entries following
+ * the current one will have the same address). If there is no exact match, the entry
+ * with the biggest address just before the given address is returned.
+ *
+ * @param symt      SYMT file header
+ * @param addr      Address to search for
+ * @param idx       If not null, will be set to the index of the entry found (or the index just before)
+ * @return          The found entry (or the entry just before)
+ */
+static addrtable_entry_t symt_addrtab_search(symtable_header_t *symt, uint32_t addr, int *idx)
+{
+    int min = 0;
+    int max = symt->addrtab_size - 1;
+    while (min < max) {
+        int mid = (min + max) / 2;
+        addrtable_entry_t entry = symt_addrtab_entry(symt, mid);
+        if (addr <= ADDRENTRY_ADDR(entry))
+            max = mid;
+        else
+            min = mid + 1;
+    }
+    addrtable_entry_t entry = symt_addrtab_entry(symt, min);
+    if (min > 0 && ADDRENTRY_ADDR(entry) > addr)
+        entry = symt_addrtab_entry(symt, --min);
+    if (idx) *idx = min;
+    return entry;
+}
+
+
+/**
+ * @brief Fetch a string from the string table
+ * 
+ * @param symt  SYMT file
+ * @param sidx  Index of the first character of the string in the string table
+ * @param slen  Length of the string
+ * @param buf   Destination buffer
+ * @param size  Size of the destination buffer
+ * @return char*  Fetched string within the destination buffer (might not be at offset 0 for alignment reasons)
+ */
+static char* symt_string(symtable_header_t *symt, int sidx, int slen, char *buf, int size)
+{
+    // Align 2-byte phase of the RAM buffer with the ROM address. This is required
+    // for dma_read.
+    int tweak = (sidx ^ (uint32_t)buf) & 1;
+    char *func = buf + tweak; size -= tweak;
+    int n = MIN(slen, size);
+
+    data_cache_hit_writeback_invalidate(buf, size);
+    dma_read(func, SYMT_ROM + symt->strtab_off + sidx, n);
+    func[n] = 0;
+    return func;
+}
+
+/**
+ * @brief Fetch a symbol table entry from the SYMT file.
+ * 
+ * @param symt    SYMT file
+ * @param entry   Output entry pointer
+ * @param idx     Index of the entry to fetch
+ */
+static void symt_entry_fetch(symtable_header_t *symt, symtable_entry_t *entry, int idx)
+{
+    data_cache_hit_writeback_invalidate(entry, sizeof(symtable_entry_t));
+    dma_read(entry, SYMT_ROM + symt->symtab_off + idx * sizeof(symtable_entry_t), sizeof(symtable_entry_t));
+}
+
+// Fetch the function name of an entry
+static char* symt_entry_func(symtable_header_t *symt, symtable_entry_t *entry, uint32_t addr, char *buf, int size)
+{
+    if (addr >= (uint32_t)inthandler && addr < (uint32_t)inthandler_end) {
+        // Special case exception handlers. This is just to show something slightly
+        // more readable instead of "notcart+0x0" or similar assembly symbols
```

Comment:
hmm, maybe we can include the subsymbols as well?

Reply:
Honestly they are just local symbols, they're not really expected to be exported. It's more of a side effect of how gas works. If you check unstable, we already changed entrypoint.S to use the syntax `.Lskip_dma` (which is how local labels should be called), so they wouldn't appear anymore in the stacktraces.

I think even if we remove all labels from stacktraces, we would end up with a single symbol called `inthandler`; as much as we could rename it, I think it does provide more clear context to the user using a special symbol like `<EXCEPTION HANDLER>`.